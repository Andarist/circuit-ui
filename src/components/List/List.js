import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'react-emotion';
import { typeMarginResets } from '../../styles/global-styles';
import { childrenPropType } from '../../util/shared-prop-types';
import HtmlElement from '../HtmlElement/HtmlElement';
import { textMega, textKilo } from '../../styles/style-helpers';
import { BIT, BYTE, KILO, MEGA } from '../../util/sizes';

const baseStyles = ({ theme }) => css`
  label: list;
  font-weight: ${theme.fontWeight.regular};
  ${typeMarginResets};
`;

const sizeStyles = ({ theme, size }) => {
  const settings = {
    [KILO]: {
      marginBottom: theme.spacings[KILO],
      marginLeft: theme.spacings[BIT],
      type: textKilo({ theme })
    },
    [MEGA]: {
      marginBottom: theme.spacings[BYTE],
      marginLeft: theme.spacings[BIT],
      type: textMega({ theme })
    }
  };
  const { marginBottom, marginLeft, type } = settings[size];
  return css`
    label: list--${size};
    ${type}
    li {
      margin-bottom: ${marginBottom};
      margin-left: ${marginLeft};
    }
    ul {
      margin-left: ${marginLeft};
    }
  `;
};

const ListElement = props => (
  <HtmlElement
    blacklist={['size', 'ordered']}
    element={({ ordered }) => (ordered ? 'ol' : 'ul')}
    {...props}
  />
);

/**
 * A list, which can be ordered or unordered
 */
const List = styled(ListElement)(baseStyles, sizeStyles);
List.KILO = KILO;
List.MEGA = MEGA;

List.propTypes = {
  /**
   * The <li> elements you want to show as a list
   */
  children: childrenPropType,
  /**
   * A Circuit UI body text size.
   */
  size: PropTypes.oneOf([KILO, MEGA]),
  /**
   * Whether the list should be presented as an <ol>
   */
  ordered: PropTypes.bool
};

List.defaultProps = {
  size: KILO,
  ordered: false
};

/**
 * @component
 */
export default List;
