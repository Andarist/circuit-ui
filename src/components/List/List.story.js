import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';

import withTests from '../../util/withTests';
import List from './List';

storiesOf('List', module)
  .addDecorator(withTests('List'))
  .add('List default', withInfo()(() =>
    <List>
      <li>This is a list</li>
      <li>A very find list</li>
      <List>
        <li>Sometimes a nested list</li>
      </List>
      <li>The kind of list you like</li>
    </List>
  ))
  .add('List ordered', withInfo()(() =>
    <List ordered>
      <li>This is a list</li>
      <li>A very find list</li>
      <List>
        <li>Sometimes a nested list</li>
      </List>
      <li>The kind of list you like</li>
    </List>
  ))
  .add('List kilo', withInfo()(() =>
    <List size="kilo">
      <li>This is a list</li>
      <li>A very find list</li>
      <List size="kilo">
        <li>Sometimes a nested list</li>
      </List>
      <li>The kind of list you like</li>
    </List>
  ))
  .add('List mega', withInfo()(() =>
    <List size="mega">
      <li>This is a list</li>
      <li>A very find list</li>
      <List size="mega">
        <li>Sometimes a nested list</li>
      </List>
      <li>The kind of list you like</li>
    </List>
  ));
