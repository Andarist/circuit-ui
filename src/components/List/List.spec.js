import React from 'react';

import List from '.';

describe('List', () => {
  it('should render a default unordered List', () => {
    const list = create(
      <List>
        <li>Hi there</li>
      </List>
    );
    expect(list).toMatchSnapshot();
  });
  it('should render a kilo unordered List', () => {
    const list = create(
      <List size={List.KILO}>
        <li>Hi there</li>
      </List>
    );
    expect(list).toMatchSnapshot();
  });
  it('should render a mega unordered List', () => {
    const list = create(
      <List size={List.MEGA}>
        <li>Hi there</li>
      </List>
    );
    expect(list).toMatchSnapshot();
  });
  it('should render nested unordered lists', () => {
    const list = create(
      <List>
        <li>Hi there</li>
        <List>
          <li>Hi there</li>
        </List>
      </List>
    );
    expect(list).toMatchSnapshot();
  });
  it('should render an ordered list', () => {
    const list = create(
      <List ordered>
        <li>Hi</li>
        <li>It's me</li>
      </List>
    );
    expect(list).toMatchSnapshot();
  });
});
