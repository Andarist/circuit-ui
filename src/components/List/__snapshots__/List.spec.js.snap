// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should render a default unordered List 1`] = `
.circuit-0 {
  font-weight: 400;
  -webkit-margin-before: 0;
  -webkit-margin-after: 0;
  -webkit-margin-start: 0;
  -webkit-margin-end: 0;
  font-size: 13px;
  line-height: 20px;
}

.circuit-0 li {
  margin-bottom: 12px;
  margin-left: 4px;
}

.circuit-0 ul {
  margin-left: 4px;
}

<ul
  className="circuit-0 circuit-1"
>
  <li>
    Hi there
  </li>
</ul>
`;

exports[`List should render a kilo unordered List 1`] = `
.circuit-0 {
  font-weight: 400;
  -webkit-margin-before: 0;
  -webkit-margin-after: 0;
  -webkit-margin-start: 0;
  -webkit-margin-end: 0;
  font-size: 13px;
  line-height: 20px;
}

.circuit-0 li {
  margin-bottom: 12px;
  margin-left: 4px;
}

.circuit-0 ul {
  margin-left: 4px;
}

<ul
  className="circuit-0 circuit-1"
>
  <li>
    Hi there
  </li>
</ul>
`;

exports[`List should render a mega unordered List 1`] = `
.circuit-0 {
  font-weight: 400;
  -webkit-margin-before: 0;
  -webkit-margin-after: 0;
  -webkit-margin-start: 0;
  -webkit-margin-end: 0;
  font-size: 15px;
  line-height: 24px;
}

.circuit-0 li {
  margin-bottom: 8px;
  margin-left: 4px;
}

.circuit-0 ul {
  margin-left: 4px;
}

<ul
  className="circuit-0 circuit-1"
>
  <li>
    Hi there
  </li>
</ul>
`;

exports[`List should render an ordered list 1`] = `
.circuit-0 {
  font-weight: 400;
  -webkit-margin-before: 0;
  -webkit-margin-after: 0;
  -webkit-margin-start: 0;
  -webkit-margin-end: 0;
  font-size: 13px;
  line-height: 20px;
}

.circuit-0 li {
  margin-bottom: 12px;
  margin-left: 4px;
}

.circuit-0 ul {
  margin-left: 4px;
}

<ol
  className="circuit-0 circuit-1"
>
  <li>
    Hi
  </li>
  <li>
    It's me
  </li>
</ol>
`;

exports[`List should render nested unordered lists 1`] = `
.circuit-0 {
  font-weight: 400;
  -webkit-margin-before: 0;
  -webkit-margin-after: 0;
  -webkit-margin-start: 0;
  -webkit-margin-end: 0;
  font-size: 13px;
  line-height: 20px;
}

.circuit-0 li {
  margin-bottom: 12px;
  margin-left: 4px;
}

.circuit-0 ul {
  margin-left: 4px;
}

<ul
  className="circuit-0 circuit-1"
>
  <li>
    Hi there
  </li>
  <ul
    className="circuit-0 circuit-1"
  >
    <li>
      Hi there
    </li>
  </ul>
</ul>
`;
